src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingScriptOperation.cs

```diff
@@ -0,0 +1,326 @@
+                publishModel.ScriptProgress += this.OnPublishModelScriptProgress;
+                publishModel.ScriptError += this.OnPublishModelScriptError;
+
+                ScriptOutputOptions outputOptions = new ScriptOutputOptions
+                {
+                    SaveFileMode = ScriptFileMode.Overwrite,
+                    SaveFileType = ScriptFileType.Unicode,
+                    SaveFileName = this.Parameters.FilePath,
+                };
+
+                publishModel.GenerateScript(outputOptions);
+
+                Logger.Write(
+                    LogLevel.Verbose,
+                    string.Format(
+                        "Sending script complete notification event with total count {0} and scripted count {1}",
+                        this.totalScriptedObjectCount,
+                        this.scriptedObjectCount));
+
+                this.SendJsonRpcEventAsync(
+                    ScriptingCompleteEvent.Type,
+                    new ScriptingCompleteParameters { OperationId = this.OperationId });
+            }
+            catch (Exception e)
+            {
+
+                if (e.IsOperationCanceledException())
+                {
+                    Logger.Write(LogLevel.Normal, string.Format("Scripting operation {0} failed was canceled", this.OperationId));
+
+                    this.SendJsonRpcEventAsync(
+                        ScriptingCancelEvent.Type,
+                        new ScriptingCancelParameters { OperationId = this.OperationId });
+                }
+                else
+                {
+                    Logger.Write(LogLevel.Error, string.Format("Scripting operation {0} failed with exception {1}", this.OperationId, e));
+
+                    ScriptingErrorParams eventParams = new ScriptingErrorParams
+                    {
+                        OperationId = this.OperationId,
+                        Message = e.Message,
+                        Details = e.ToString(),
+                    };
+
+                    this.SendJsonRpcEventAsync(ScriptingErrorEvent.Type, eventParams);
+                }
+            }
+            finally
+            {
+                if (publishModel != null)
+                {
+                    publishModel.ScriptItemsCollected -= this.OnPublishModelScriptItemsCollected;
+                    publishModel.ScriptProgress -= this.OnPublishModelScriptProgress;
+                    publishModel.ScriptError -= this.OnPublishModelScriptError;
+                }
+            }
+        }
+
+        private SqlScriptPublishModel BuildPublishModel()
+        {
+            SqlScriptPublishModel publishModel = new SqlScriptPublishModel(this.Parameters.ConnectionString);
+            PopulateAdvancedScriptOptions(publishModel.AdvancedOptions);
+
+            bool hasIncludeCriteria = this.Parameters.IncludeObjectCriteria != null && this.Parameters.IncludeObjectCriteria.Any();
```

Comment:
I think it's readable the way it is and I'm okay with multiple null/Any calls on the collection. The key optimization begin made is avoiding the call to ScriptingObjectMatchProcessor.Match if possible since it needs to load all of the objects from the database. I added a comment to make that more explicit.